sophia-fungible-token-contract.md

```diff
@@ -0,0 +1,491 @@
+  let c : int = a + b
+  require(c >= a, "Error")
+  c
+```
+- **sub** - subtracts two unsigned integers and return the result, throws on overflow (i.e. if subtrahend is greater than minuend);
+```
+private function sub(a : int, b : int) : int =
+  require(b =< a, "Error")
+  a - b
+```
+
+- **onlyOwner** - throws if called by any account other than the owner
+```
+private function onlyOwner() =
+  require(Call.caller == state.owner, "Only owner can mint!")
+```
+
+- **lookupByAddress** - checks if a specific address (the first parameter) exists in second parameter of type map, if it doesn't the function returns the third parameter passed to it;
+```
+private function lookupByAddress(k : address, m, v) =
+  switch(Map.lookup(k, m))
+    None    => v
+    Some(x) => x
+```
+
+### Implementation of the fungible token
+
+We will write a number of functions which will allow us to find out the balances of accounts as well as to transfer them from one account to another under varying conditions. These functions are described below:
+
+- **balanceOf** - the function provides the number of tokens held by a given address. ***Note*** that anyone can query any address’ balance, as all data on the blockchain is public;
+
+```
+public function balanceOf(who: address) : int = lookupByAddress(who, state.balances, 0)
+```
+
+- **totalSupply** - the return value equals the sum of all balances;
+```
+public function totalSupply() : int = state.totalSupply
+```
+
+- **transfer** - the function transfers a number of tokens directly from the caller to another address;
+```
+public stateful function transfer(to: address, value: int) : bool =
+  _transfer(Call.caller, to, value)
+    
+private stateful function _transfer(from: address, to: address, value: int) : bool =
+  require(value > 0, "Value is sub zero")
+  require(value =< balanceOf(from), "Not enough balance")
+  require(to != #0, "Invalid address")
+
+  put(state{
+    balances[from] = sub(balanceOf(from), value),
+    balances[to] = add(balanceOf(to), value)})
+
+  true
+```
+- **transferFrom** and **approve** - two functions that allow to transfer a number of tokens using a two-step process. In the first step a token holder gives another address approval to transfer up to a certain number of tokens from his balance to an address of their choice. This approved amount is known as an allowance. The token holder uses approve() to provide this information;
+```
+public stateful function approve(spender: address, value: int) : bool =
+  require(value > 0, "Value is sub zero")
+  require(spender != #0, "Invalid spender address")
+
+  put(state{allowed[(Call.caller,spender)] = value})
+
+  true
+```
+
+```
+public stateful function transferFrom(from: address, to: address, value: int) : bool =
+  require(state.allowed[(from, Call.caller)] >= value, "Value is bigger than allowed")
+
+  put(state{allowed[(from, Call.caller)] = sub(state.allowed[(from, Call.caller)], value)})
+  _transfer(from, to, value)
```

Comment:
this notion should be either `transfer` or if already defined with the same name as parent function with `transfer'`